### ISSUE TITLE
[expo-linking] `createURL` double-encodes URIs

### ISSUE THREAD

**Issue by billinghamj** [expo/expo#22891]
### Minimal reproducible example

https://github.com/expo/expo/pull/7934#discussion_r414922066

### Summary

It looks like expo-linking has had a bug with `createURL` double-encoding URIs from when it was first introduced.

In Expo's own internal code review, the issue was identified, but it looks like it was just never fixed: https://github.com/expo/expo/pull/7934#discussion_r414922066

e.g.:

```ts
Linking.createURL('/redirect', {
  isTripleSlashed: true,
  queryParams: { path: '/a/b/c' },
});
```

- Should return: `foo:///redirect?path=%2Fa%2Fb%2Fc`
- Actually returns: `foo:///redirect?path=%252Fa%252Fb%252Fc` (path is incorrectly double encoded)

This is because `encodeURI` is called when it shouldn't be (in both the web & non-web code paths)

---

### Compatibility

Right now, you can reliably just call `decodeURI` on `createURL`'s output, and you then get a valid result.

As this behaviour has been there for so long, I'm not too clear on how you'd typically handle something like this. It seems like a potentially serious breaking change, perhaps serious enough to never change the behaviour of this specific method (e.g. instead replacing the method name, or documenting the weird behaviour)

The reason I ask is because I need to know whether I can rely on being able to use `decodeURI`, or whether it's likely to be fixed with this method name, and applying that workaround could break in the next SDK version? (And therefore maybe we should generate the query params in our own code)

### Environment

```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 14.0
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.1.0 - /opt/homebrew/bin/node
      Yarn: 1.22.19 - /opt/homebrew/bin/yarn
      npm: 9.6.4 - /opt/homebrew/bin/npm
    Managers:
      CocoaPods: 1.12.1 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 3.13.0
```

**Comment by billinghamj** [comment]
This is an active issue, just nobody has looked at it 🙄

**Comment by billinghamj** [comment]
Not stale

**Comment by billinghamj** [comment]
Not stale :|

**Comment by billinghamj** [comment]
Nope

**Comment by nucleartux** [comment]
Might be related, but I'm not sure: https://github.com/facebook/react-native/issues/39793

**Comment by billinghamj** [comment]
@nucleartux I don't think so - the code in expo-linking is just not correct, their own PR review identified the issue before merge

**Comment by billinghamj** [comment]
Not stale

**Comment by billinghamj** [comment]
Not stale

**Comment by notmyrealcode** [comment]
I can reproduce as well.

**Comment by billinghamj** [comment]
It isn't stale
